### PR TITLE
Support a debug framework for TVM Runtime

### PR DESCRIPTION
**Pull request by siju-samuel**
A debug framework for TVM Runtime.

Refer [this link](https://github.com/siju-samuel/tvm/blob/tvmdbg/docs/dev/tvmdbg.md
) for the user manual.

**This PR supports the below features.**
  - [x] Show fused graph summary
  - [x] Perform debug run and show node details(inputs/outputs)
  - [x] Run without debug
**The below features I'm still  working on and will PR later**
  - [ ] Call graph run-time n times from UI
  - [ ] Support check for NAN in data and break
  - [ ] Support check for INF in data and break
  - [ ] Support for step debug(debug node by node)
  - [ ] Support changing node input values and running a specific node
  - [ ] Support inject node input values from a dump file
  - [ ] Support dumping of node outputs to a file
  - [ ] Support comparison of node output with a dump output
  - [ ] Support profiler for performance debugging
  - [ ] Test framework for tvmdbg

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for contributing to this. Because this is a big change to the framework, I would recommend starting an RFC issue first, so that various design aspects can be discussed among the contributors, before we start reviewing the PR.

**Comment by tqchen** [comment]
One thing we want to avoid is to make sure that the image files are not stored in this repo, to minimize the size of the history

**Comment by siju-samuel** [comment]
https://github.com/dmlc/tvm/issues/1315

File: python/tvm/tools/debug/ui/curses_widgets.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+NAVIGATION_MENU_COLOR_ATTR = "white_on_black"
+
+class NavigationHistoryItem(object):
+    """Individual item in navigation history."""
+
+    def __init__(self, command, screen_output, scroll_position):
+        """Constructor of NavigationHistoryItem.
+
+        Args:
+          command: (`str`) the command line text.
```
Comment (merrymercy) [inline python/tvm/tools/debug/ui/curses_widgets.py]: Should use this document style: https://docs.tvm.ai/contribute/document.html#document-python

File: src/runtime/graph/graph_runtime.cc [hunk 2]
```diff
@@ -56,6 +57,25 @@ class GraphRuntime : public ModuleNode {
...
+      if (op_execs_[i]) op_execs_[i]();
+      gettimeofday(&tp, NULL);
+      int64_t mstime = int64_t(tp.tv_sec * 1000000L + tp.tv_usec);
+      size_t num_outputs = (nodes_[i].op_type == "null") ? 1: nodes_[i].param.num_outputs;
+      for (size_t j = 0; j < num_outputs; j++) {
+          uint32_t eid = this->entry_id(i, j);
+          TVM_CCALL(TVMArrayCopyFromTo(&data_entry_[eid],
+                                       &debug_buffers_[eid]->out_tensor,
+                                       nullptr));
+          debug_buffers_[eid]->time_stamp = mstime;
```
Comment (merrymercy) [inline src/runtime/graph/graph_runtime.cc]: I prefer to use layer-wise time cost to replace timestamp here.
To say, set timer to only record the time cost of 
`if (op_execs_[i]) op_execs_[i]();`

The way you record time stamp also includes the overhead of data copy.  These time stamps are not identical to the real time stamps when debug mode is off.

**Comment by tqchen** [comment]
this is closed for now, once the RFC discussion is finished and we have actionable items, we can reopen this PR to do the review

**Pull request closed without merging**